platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -425,93 +456,198 @@ public DataFileReader endWriting() throws IOException {
+                // tasks created is limited by the size of oneTransactionsData, which is not
+                // expected to be huge, since data flushes are limited
+                while (it.hasNext()) {
+                    IntObjectPair<BucketMutation> keyValue = it.next();
                     final int bucketIndex = keyValue.getOne();
                     final BucketMutation bucketMap = keyValue.getTwo();
-                    flushExecutor.execute(() -> readUpdateQueueBucket(bucketIndex, bucketMap, queue));
-                    ++inFlight;
+                    // Create a "read bucket" task. Executed, once its out is set
+                    final ReadBucketTask readBucketTask = new ReadBucketTask(pool, bucketIndex);
+                    // Create an "update bucket" task. It's the out for the previous task. Executed
+                    // after the bucket is read by the read task and set as input for this task
+                    final UpdateBucketTask updateBucketTask = new UpdateBucketTask(pool, bucketMap);
+                    // This will trigger read bucket task execution
+                    readBucketTask.setOut(updateBucketTask);
                 }
-
-                ReadBucketResult res;
-                while ((res = queue.poll()) != null) {
-                    --inFlight;
-                    if (res.error != null) {
-                        throw new RuntimeException(res.error);
-                    }
-                    try (final Bucket bucket = res.bucket) {
-                        final int bucketIndex = bucket.getBucketIndex();
-                        if (bucket.isEmpty()) {
-                            // bucket is missing or empty, remove it from the index
-                            bucketIndexToBucketLocation.remove(bucketIndex);
-                        } else {
-                            // save bucket
-                            final long bucketLocation =
-                                    fileCollection.storeDataItem(bucket::writeTo, bucket.sizeInBytes());
-                            // update bucketIndexToBucketLocation
-                            bucketIndexToBucketLocation.put(bucketIndex, bucketLocation);
-                        }
-                    } finally {
-                        ++processed;
-                    }
+                // Wait until all tasks are completed. The latch is set in the end of the last
+                // executed "store bucket" task or when an exception is occurred
+                finishWritingLatch.await();
+                if (exceptionOccurred.get() != null) {
+                    throw exceptionOccurred.get();
                 }
+                // close files session
+                dataFileReader = fileCollection.endWriting(0, numOfBuckets);
+                // we have updated all indexes so the data file can now be included in merges
+                dataFileReader.setFileCompleted();
```

Comment:
Minor: I'm unsure about the logic that calls this method, but these finalization tasks won't be executed if an exception is thrown above. Should these go to a `finally{}` block?

Reply:
If an exception is thrown, it isn't recoverable, it doesn't make sense to pretend that the file has been written successfully. This is no different from the current implementation.

Reply:
Sounds good as long as this doesn't lead to files corruption on disk.